File: game/render/combat.lua
---@class CombatRenderModule
local combat_render = {}

local FAMILY_COLORS = {
    unarmed = { 0.55, 0.95, 1 },
    guard = { 0.75, 0.9, 1 },
    light_melee = { 1, 0.72, 0.28 },
    ranged = { 1, 0.45, 0.78 },
}

---@param direction Vec2
---@return number
---@return number
local function unit(direction)
    local length = math.sqrt(direction.x * direction.x + direction.y * direction.y)
    if length < 1e-9 then
        return 1, 0
    end
    return direction.x / length, direction.y / length
end

---@param sample CombatPlayerPresentation
---@return number
local function telegraph_alpha(sample)
    if sample.phase == "active" then
        return 0.95
    elseif sample.phase == "guard" or sample.phase == "aim" then
        return 0.72
    end
    return 0.35 + sample.phase_progress * 0.35
end

---@param project fun(wx: number, wy: number): number, number, number
---@param sample CombatPlayerPresentation
---@param position { x: number, y: number }
local function draw_arc(project, sample, position)
    local color = assert(FAMILY_COLORS[assert(sample.family_id)])
    local direction_x, direction_y = unit(sample.direction)
    local heading = math.atan2(direction_y, direction_x)
    local half_arc = math.rad(assert(sample.front_arc_degrees) / 2)
    local radius = sample.reach_px or 34
    local alpha = telegraph_alpha(sample)
    local center_x, center_y = project(position.x, position.y)
    local prior_x, prior_y ---@type number?, number?
    local first_x, first_y ---@type number?, number?
    local last_x, last_y ---@type number?, number?
    love.graphics.setColor(color[1], color[2], color[3], alpha)
    love.graphics.setLineWidth(sample.telegraph_kind == "guard_arc" and 2.5 or 1.8)
    for index = 0, 14 do
        local angle = heading - half_arc + 2 * half_arc * index / 14
        local world_x = position.x + math.cos(angle) * radius
        local world_y = position.y + math.sin(angle) * radius
        local screen_x, screen_y = project(world_x, world_y)
        if prior_x then
            love.graphics.line(prior_x, assert(prior_y), screen_x, screen_y)
        else
            first_x, first_y = screen_x, screen_y
        end
        prior_x, prior_y = screen_x, screen_y
        last_x, last_y = screen_x, screen_y
    end

    if sample.telegraph_kind == "guard_arc" then
        local inner_prior_x, inner_prior_y ---@type number?, number?
        for index = 0, 10 do
            local angle = heading - half_arc + 2 * half_arc * index / 10
            local screen_x, screen_y = project(
                position.x + math.cos(angle) * radius * 0.7,
                position.y + math.sin(angle) * radius * 0.7
            )
            if inner_prior_x then
                love.graphics.line(inner_prior_x, assert(inner_prior_y), screen_x, screen_y)
            end
            inner_prior_x, inner_prior_y = screen_x, screen_y
        end
        love.graphics.line(center_x, center_y, assert(first_x), assert(first_y))
        love.graphics.line(center_x, center_y, assert(last_x), assert(last_y))
    else
        local tip_x, tip_y =
            project(position.x + direction_x * radius, position.y + direction_y * radius)
        love.graphics.line(center_x, center_y, tip_x, tip_y)
        if sample.family_id == "unarmed" then
            love.graphics.circle("line", tip_x, tip_y, 4)
        else
            local normal_x, normal_y = -direction_y, direction_x
            local left_x, left_y = project(
                position.x + direction_x * radius + normal_x * 7,
                position.y + direction_y * radius + normal_y * 7
            )
            local right_x, right_y = project(
                position.x + direction_x * radius - normal_x * 7,
                position.y + direction_y * radius - normal_y * 7
            )
            love.graphics.line(left_x, left_y, right_x, right_y)
        end
    end
    love.graphics.setLineWidth(1)
end

---@param project fun(wx: number, wy: number): number, number, number
---@param sample CombatPlayerPresentation
---@param position { x: number, y: number }
local function draw_line(project, sample, position)
    local color = FAMILY_COLORS.ranged
    local direction_x, direction_y = unit(sample.direction)
    local range = sample.projectile_range_px or 300
    local alpha = telegraph_alpha(sample)
    local start_x, start_y = project(position.x, position.y)
    local end_x, end_y = project(position.x + direction_x * range, position.y + direction_y * range)
    love.graphics.setColor(color[1], color[2], color[3], alpha)
    love.graphics.setLineWidth(sample.phase == "active" and 3 or 1.5)
    love.graphics.line(start_x, start_y, end_x, end_y)

    -- Repeated transverse gates make range and direction legible without color.
    local normal_x, normal_y = -direction_y, direction_x
    for step = 1, 4 do
        local distance = range * step / 5
        local left_x, left_y = project(
            position.x + direction_x * distance + normal_x * 5,
            position.y + direction_y * distance + normal_y * 5
        )
        local right_x, right_y = project(
            position.x + direction_x * distance - normal_x * 5,
            position.y + direction_y * distance - normal_y * 5
        )
        love.graphics.line(left_x, left_y, right_x, right_y)
    end
    love.graphics.setLineWidth(1)
end

---@param model CombatPresentationModel
---@param project fun(wx: number, wy: number): number, number, number
---@param pose CorrectionSmoothingPose?
function combat_render.draw_under(model, project, pose)
    if not model.enabled then
        return
    end
    for _, sample in ipairs(model.players) do
        local position = (pose and pose.players[sample.player_id]) or sample.position
        if sample.telegraph_kind == "line" then
            draw_line(project, sample, position)
        elseif sample.telegraph_kind then
            draw_arc(project, sample, position)
        end
    end
end

---@param model CombatPresentationModel
---@param project fun(wx: number, wy: number): number, number, number
function combat_render.draw_over(model, project)
    if not model.enabled then
        return
    end
    for _, projectile in ipairs(model.projectiles) do
        local direction_x, direction_y = unit(projectile.direction)
        local tail_x, tail_y = project(
            projectile.position.x - direction_x * 18,
            projectile.position.y - direction_y * 18
        )
        local x, y, scale = project(projectile.position.x, projectile.position.y)
        local radius = math.max(3, 5 * scale)
        love.graphics.setColor(1, 0.45, 0.78, 0.58)
        love.graphics.setLineWidth(math.max(1, 3 * scale))
        love.graphics.line(tail_x, tail_y, x, y)
        love.graphics.setColor(1, 0.9, 1, 0.98)
        love.graphics.polygon("fill", x, y - radius, x + radius, y, x, y + radius, x - radius, y)
        love.graphics.setColor(0.15, 0.03, 0.18, 0.9)
        love.graphics.circle("fill", x, y, radius * 0.32)
    end
    love.graphics.setLineWidth(1)
end

return combat_render

File: v2/ts/packages/render/src/combat.ts
// Ported from game/render/combat.lua.
//
// Combat telegraphs (arcs, guard fans, ranged sightlines) and projectile
// trails, drawn on the ground under the players (`drawUnderCommands`) and
// on top of everything (`drawOverCommands`). Pure content: every function
// here returns a `DrawCommand[]` (draw2d.ts) rather than calling
// `love.graphics`/three.js directly, so it is fully testable headless.
//
// Boundary note (v2/README.md rule 6.7): `CombatPlayerPresentation` and
// `CombatProjectilePresentation` are `@gc/presentation`'s (already a
// declared dependency), reused rather than redeclared. `RenderFrame` is the
// (Rust) `render/frame.lua` producer's output (`render/**` -> Rust
// `crates/gc-render`, v2/README.md #2) -- only the slice this module reads
// is declared locally. `player_index` on a `CombatPlayerPresentation` is
// 1-based (an identifier that survives the wasm boundary, matching
// `@gc/presentation`'s own convention), so it is adjusted by one when
// indexing this module's 0-based `RenderFrame.players` arrays.

import type { Vec2 } from "@gc/core";
import type { CombatPlayerPresentation, CombatProjectilePresentation } from "@gc/presentation";
import { DrawList, paint, type DrawCommand, type Project, type RGB } from "./draw2d.ts";
import * as THREE from "three";

/** The slice of `RenderFrame` this module reads. */
export interface CombatDrawFrame {
  readonly combat?: {
    readonly enabled: boolean;
    readonly players: readonly CombatPlayerPresentation[];
    readonly projectiles: readonly CombatProjectilePresentation[];
  };
  readonly players: {
    readonly x: readonly number[];
    readonly y: readonly number[];
  };
}

const FAMILY_COLORS: Readonly<Record<string, RGB>> = {
  unarmed: [0.55, 0.95, 1],
  guard: [0.75, 0.9, 1],
  light_melee: [1, 0.72, 0.28],
  ranged: [1, 0.45, 0.78],
};

function familyColor(familyId: string | undefined): RGB {
  const color = familyId !== undefined ? FAMILY_COLORS[familyId] : undefined;
  if (color === undefined) {
    throw new Error(`combat.ts: unknown or missing family_id: ${String(familyId)}`);
  }
  return color;
}

function unit(direction: Vec2): readonly [number, number] {
  const length = Math.sqrt(direction.x * direction.x + direction.y * direction.y);
  if (length < 1e-9) {
    return [1, 0];
  }
  return [direction.x / length, direction.y / length];
}

function telegraphAlpha(sample: CombatPlayerPresentation): number {
  if (sample.phase === "active") {
    return 0.95;
  } else if (sample.phase === "guard" || sample.phase === "aim") {
    return 0.72;
  }
  return 0.35 + sample.phase_progress * 0.35;
}

function drawArc(dl: DrawList, project: Project, sample: CombatPlayerPresentation, position: { x: number; y: number }): void {
  const color = familyColor(sample.family_id);
  const [directionX, directionY] = unit(sample.direction);
  const heading = Math.atan2(directionY, directionX);
  const frontArcDegrees = sample.front_arc_degrees;
  if (frontArcDegrees === undefined) {
    throw new Error("combat.ts: draw_arc requires front_arc_degrees");
  }
  const halfArc = (frontArcDegrees / 2) * (Math.PI / 180);
  const radius = sample.reach_px ?? 34;
  const alpha = telegraphAlpha(sample);
  const [centerX, centerY] = project(position.x, position.y);
  const lineWidth = sample.telegraph_kind === "guard_arc" ? 2.5 : 1.8;

  let priorX: number | undefined;
  let priorY: number | undefined;
  let firstX: number | undefined;
  let firstY: number | undefined;
  let lastX: number | undefined;
  let lastY: number | undefined;
  for (let index = 0; index <= 14; index += 1) {
    const angle = heading - halfArc + (2 * halfArc * index) / 14;
    const worldX = position.x + Math.cos(angle) * radius;
    const worldY = position.y + Math.sin(angle) * radius;
    const [screenX, screenY] = project(worldX, worldY);
    if (priorX !== undefined && priorY !== undefined) {
      dl.line([priorX, priorY, screenX, screenY], color, { alpha, lineWidth });
    } else {
      firstX = screenX;
      firstY = screenY;
    }
    priorX = screenX;
    priorY = screenY;
    lastX = screenX;
    lastY = screenY;
  }

  if (sample.telegraph_kind === "guard_arc") {
    let innerPriorX: number | undefined;
    let innerPriorY: number | undefined;
    for (let index = 0; index <= 10; index += 1) {
      const angle = heading - halfArc + (2 * halfArc * index) / 10;
      const [screenX, screenY] = project(position.x + Math.cos(angle) * radius * 0.7, position.y + Math.sin(angle) * radius * 0.7);
      if (innerPriorX !== undefined && innerPriorY !== undefined) {
        dl.line([innerPriorX, innerPriorY, screenX, screenY], color, { alpha, lineWidth });
      }
      innerPriorX = screenX;
      innerPriorY = screenY;
    }
    if (firstX === undefined || firstY === undefined || lastX === undefined || lastY === undefined) {
      throw new Error("combat.ts: draw_arc guard fan requires at least one arc sample");
    }
    dl.line([centerX, centerY, firstX, firstY], color, { alpha, lineWidth });
    dl.line([centerX, centerY, lastX, lastY], color, { alpha, lineWidth });
  } else {
    const [tipX, tipY] = project(position.x + directionX * radius, position.y + directionY * radius);
    dl.line([centerX, centerY, tipX, tipY], color, { alpha, lineWidth });
    if (sample.family_id === "unarmed") {
      dl.circle("line", tipX, tipY, 4, color, { alpha, lineWidth });
    } else {
      const normalX = -directionY;
      const normalY = directionX;
      const [leftX, leftY] = project(position.x + directionX * radius + normalX * 7, position.y + directionY * radius + normalY * 7);
      const [rightX, rightY] = project(position.x + directionX * radius - normalX * 7, position.y + directionY * radius - normalY * 7);
      dl.line([leftX, leftY, rightX, rightY], color, { alpha, lineWidth });
    }
  }
}

function drawLine(dl: DrawList, project: Project, sample: CombatPlayerPresentation, position: { x: number; y: number }): void {
  const color = familyColor("ranged");
  const [directionX, directionY] = unit(sample.direction);
  const range = sample.projectile_range_px ?? 300;
  const alpha = telegraphAlpha(sample);
  const [startX, startY] = project(position.x, position.y);
  const [endX, endY] = project(position.x + directionX * range, position.y + directionY * range);
  const lineWidth = sample.phase === "active" ? 3 : 1.5;
  dl.line([startX, startY, endX, endY], color, { alpha, lineWidth });

  const normalX = -directionY;
  const normalY = directionX;
  for (let step = 1; step <= 4; step += 1) {
    const distance = (range * step) / 5;
    const [leftX, leftY] = project(position.x + directionX * distance + normalX * 5, position.y + directionY * distance + normalY * 5);
    const [rightX, rightY] = project(position.x + directionX * distance - normalX * 5, position.y + directionY * distance - normalY * 5);
    dl.line([leftX, leftY, rightX, rightY], color, { alpha, lineWidth });
  }
}

/** Telegraphs on the ground under the players. See file header. */
export function drawUnderCommands(frame: CombatDrawFrame, project: Project): DrawCommand[] {
  const dl = new DrawList();
  const model = frame.combat;
  if (model === undefined || !model.enabled) {
    return dl.commands;
  }
  for (const sample of model.players) {
    const slot = sample.player_index - 1; // 1-based content id -> 0-based array index
    const x = frame.players.x[slot];
    const y = frame.players.y[slot];
    if (x === undefined || y === undefined) {
      continue;
    }
    const position = { x, y };
    if (sample.telegraph_kind === "line") {
      drawLine(dl, project, sample, position);
    } else if (sample.telegraph_kind !== undefined) {
      drawArc(dl, project, sample, position);
    }
  }
  return dl.commands;
}

/** Projectile trails, drawn on top of everything. See file header. */
export function drawOverCommands(frame: CombatDrawFrame, project: Project): DrawCommand[] {
  const dl = new DrawList();
  const model = frame.combat;
  if (model === undefined || !model.enabled) {
    return dl.commands;
  }
  for (const projectile of model.projectiles) {
    const [directionX, directionY] = unit(projectile.direction);
    const [tailX, tailY] = project(projectile.position.x - directionX * 18, projectile.position.y - directionY * 18);
    const [x, y, scale] = project(projectile.position.x, projectile.position.y);
    const radius = Math.max(3, 5 * scale);
    dl.line([tailX, tailY, x, y], [1, 0.45, 0.78], { alpha: 0.58, lineWidth: Math.max(1, 3 * scale) });
    dl.polygon("fill", [x, y - radius, x + radius, y, x, y + radius, x - radius, y], [1, 0.9, 1], { alpha: 0.98 });
    dl.circle("fill", x, y, radius * 0.32, [0.15, 0.03, 0.18], { alpha: 0.9 });
  }
  return dl.commands;
}

/** Impure: paints the under-players telegraphs into `group`. Untested -- see draw2d.ts. */
export function drawUnder(group: THREE.Group, frame: CombatDrawFrame, project: Project): void {
  paint(group, drawUnderCommands(frame, project));
}

/** Impure: paints the over-everything projectile trails into `group`. Untested -- see draw2d.ts. */
export function drawOver(group: THREE.Group, frame: CombatDrawFrame, project: Project): void {
  paint(group, drawOverCommands(frame, project));
}
